server/ngwmn/services/ngwmn.py

```diff
@@ -188,83 +188,147 @@ def get_well_log(agency_cd, location_id):
+                    'ui': (lambda words: {
+                        'colors': get_colors(words),
+                        'materials': classify_material(words)
+                    })(re.findall(r'\w+', _find(unit, 'gml:description', default='').lower())),
+                    'purpose': _find(unit, 'gsml:purpose')
+                }
+
+                part = unit.find('gsml:composition/gsml:CompositionPart', xml.nsmap)
+                if part is not None:
+                    log_entry_unit_composition = {
+                        'role': _find(part, 'gsml:role')
+                    }
+
+                    lith = part.find('gsml:lithology/gsml:ControlledConcept/gml:name', xml.nsmap)
+                    if lith is not None:
+                        log_entry_unit_composition['lithology'] = {
+                            'scheme': lith.get('codeSpace'),
+                            'value': lith.text
+                        }
+
+                    material = part.find('gsml:material/gsml:UnconsolidatedMaterial', xml.nsmap)
+                    if material is not None:
+                        log_entry_unit_composition['material'] = {
+                            'name': _find(material, 'gml:name'),
+                            'purpose': _find(material, 'gsml:purpose')
+                        }
+
+                    proportion = part.find('gsml:proportion/gsml:CGI_TermValue/gsml:value', xml.nsmap)
+                    if proportion is not None:
+                        log_entry_unit_composition['proportion'] = {
+                            'scheme': proportion.get('codeSpace'),
+                            'value': proportion.text
+                        }
+
+                    log_entry['unit']['composition'] = log_entry_unit_composition
+
+            shape = entry.find('gsml:shape/gml:LineString', xml.nsmap)
+            if shape is not None:
+                log_entry['shape'] = {
+                    'dimension': shape.get('srsDimension'),
+                    'unit': _default(shape.get('uom'), 'ft'),
+                    'coordinates': _coordinates(_find(shape, 'gml:coordinates'))
+                }
+
+            result['log_entries'].append(log_entry)
+
+    construction = []
+
+    casings = water_well.findall('gwml:construction/gwml:WellCasing/gwml:wellCasingElement/gwml:WellCasingComponent', xml.nsmap)
+    for index, elem in enumerate(casings):
+        casing = {
             'id': f'casing-{index}',
             'type': 'casing',
-            'position': (lambda line: {
+            'material': _find(elem, 'gwml:material/gsml:CGI_TermValue/gsml:value')
+        }
+
+        line = elem.find('gwml:position/gml:LineString', xml.nsmap)
+        if line is not None:
+            casing['position'] = {
                 'unit': _default(_find(line, 'gml:uom'), 'ft'),
                 'coordinates': _coordinates(_find(line, 'gml:coordinates'))
-            })(elem.find('gwml:position/gml:LineString', xml.nsmap)),
```

Comment:
To preserve the declarative aspect of this function, it'd also be possible to do the null check inline like:
```python
} if line else None)(elem.find('gwml:position/gml:LineString', xml.nsmap)),
```

Reply:
I thought of that too. I think making this code less declarative makes sense at this time. I was having trouble determining where issues were arising and since the incoming data is very dirty, I think this will help track down issues. YMMV